### DISCUSSION TITLE
how use exclusive implements leader election ??i want know more details??

### DISCUSSION THREAD

**Discussion by startjava** [apache/pulsar#17703] [Q&A] [answered]
@merlimat 

https://github.com/apache/pulsar/wiki/PIP-68%3A-Exclusive-Producer

There are several use cases that require exclusive access for a single writer, of which few examples are:
(1)Ensuring a linear non-interleaved history of messages
(2)Providing basic mechanism for leader election

how use exclusive implements leader election ??i want know more details??

thank you very much!

**Comment by startjava** [comment]
about :
One example of concrete use case for this feature is in the Pulsar Functions metadata controller. In order to write a single linear history of all the functions metadata updates, it requires to elect 1 leader and that all the "decisions" made by this leader be written on the topic.

**Comment by merlimat** [comment, marked as answer]
@startjava You can implement a "leader election" pattern using the "exclusive producer" feature. 

Basically, if a certain producer is able to become the "exclusive producer" on the topic, it will assume itself the role of "leader". It is important though that all the decisions taken by the leader to be communicated as messages published in the topic.

There are few options to chose from when creating the exclusive producer: 
 * `Exclusive`: Require exclusive access for producer. Gives error if there's already a producer
 * `ExclusiveWithFencing`: Acquire exclusive access for the producer. - Removes any existing producers
 * `Shared`: By default multiple producers can publish on a topic.
 * `WaitForExclusive`: Producer creation is pending until it can acquire exclusive access.
